### ISSUE TITLE
Enhancement: [return-await] change rule option to object

### ISSUE THREAD

**Issue by phaux** [typescript-eslint/typescript-eslint#9546]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/return-await/

### Description

From https://github.com/typescript-eslint/typescript-eslint/issues/9030#issuecomment-2083721143

> Another approach here might be restructuring the options as  
> ```
> "affectsControlFlow": "always" | "never" | "dontCare",  
> "doesNotAffectControlFlow": "always" | "never" | "dontCare"  
> ```
> but that seems needlessly breaking 🤷

and https://github.com/typescript-eslint/typescript-eslint/pull/9364#issuecomment-2211750536

> Maybe change the option so that instead of a string it's an object:
> 
> ```ts
> {
>   default: "always" | "never" | "ignore", 
>   inTryCatch?: "always" | "never" | "ignore", // undefined means "same as default" 
> }
> ```
> 
> And make the string options aliases for backward compatibility:
> 
> -   "always" - `{ default: "always" }`
> -   "in-try-catch" - `{ default: "never", inTryCatch: "always" }`
> -   "never" - `{ default: "never" }`
> 
> The new option would be `{ default: "ignore", inTryCatch: "always" }`
> 
> In the future, there could be more conditions added. For example, someone might want to use "always" but make an exception for single-line arrow functions to make them less noisy:
> 
> ```ts
> /* eslint @typescript-eslint/promise-function-async: "error" */
> /* eslint @typescript-eslint/return-await: ["error", { default: "always", inArrowFunction: "ignore" }] */
> 
> promise.then(async (v) => doSomethingAsync(v)) // allowed
> ```

### Fail

```TypeScript
// no changes
```

### Pass

```TypeScript
// no changes
```

**Comment by phaux** [comment]
Reply to @Josh-Cena 

> I thought about the object API (which I mentioned in the linked issue), but I now don't like it. It creates a lot of "dubious combinations", for example you would never want `{ default: "always", inTryCatch: "never" }`.

Then remove the possibilities that don't make sense from the types (assuming "never" is deprecated):

```ts
type Options = "always" | {
  default: "always" | "ignore" | "never",
  // inTryCatch: "always", // not even necessary because you can't disable it
  inArrowFunctionShorthand?: "ignore" | "never",
}
```

> > In the future, there could be more conditions added. For example, someone might want to use "always" but make an exception for single-line arrow functions to make them less noisy:
> 
> You should either not mark it as async (which is even less noisy), or add the await. I don't think we should accommodate that style as that's not really the spirit of this rule.

This was just an example of how it would look if the rule had to be extended again.

But if we assume we never want another modifier then all we are left with is just:

```ts
type Options = 
  | "always"
  | { default: "ignore", /* inTryCatch: "always" */ }
  | { default: "never", /* inTryCatch: "always" */ }
```

Which could be simplified to just `"always" | "prefer-ignore | "prefer-never"`.
@kirkwaiblinger maybe use these names in your PR?

**Comment by kirkwaiblinger** [comment]
TLDR; IMO we should _not_ change the rule structure unless/until there are other concrete options added to the rule.

***

The way I, personally, see the rule for now is that, after #9364, it should _always_ enforce `return await` in error handling scenarios.

We have two stylistic choices 
- prefer `return await promise` ("always")
- prefer `return promise` ("in-try-catch")

As for the other options, we have 
- a stylistically unopinionated config which is a concession in order to be able to add the rule to a preset ("error-handling-correctness-only"), but not something we actually want people to use.
- A deprecated, IMO harmful option, which we can omit from this conversation ("never")

So, that gives us three options states that we need to think about. I do not believe that there is a benefit to restructuring to an options object to support just those options.

_If_ a concrete proposal for an option is created and accepted, I think that switching to an object-based options structure (with aliases to avoid breaking changes) may well be a viable solution. But IMO that needs to exist first, before it's worth considering this issue.

**Comment by kirkwaiblinger** [comment]
(I acknowledge that my current position may appear at odds with some of my previous comments referenced in the issue report. The context around why I considered that idea at the time has changed, and so has my opinion. I don't mean to pull the rug out from under @phaux 🙂)

**Comment by Josh-Cena** [comment]
I agree. The current three choices: always, in-try-catch, in-try-catch-only (?) correspond to `default: "always" | "never" | "ignore"`, but the whole premise of this proposal is that people want to handle concise arrows differently, and that needs to be proposed separately, as I don't think that's a nice convention to enforce.

**Comment by kirkwaiblinger** [comment]
Closing for now, but I would invite this to be reopened/refiled if a concrete option proposal is created as discussed.
